src/GraphQL.MicrosoftDI/GraphQLBuilder.cs

```diff
@@ -0,0 +1,126 @@
+    {
+        private readonly IServiceCollection _services;
+        /// <summary>
+        /// Initializes a new instance for the specified service collection.
+        /// </summary>
+        /// <remarks>
+        /// Registers various default services via <see cref="GraphQLBuilderBase.Initialize"/>.
+        /// </remarks>
+        public GraphQLBuilder(IServiceCollection services)
+        {
+            _services = services ?? throw new ArgumentNullException(nameof(services));
+            services.AddOptions();
+            Initialize();
+        }
+
+        private static MSServiceLifetime TranslateLifetime(ServiceLifetime serviceLifetime)
+            => serviceLifetime switch
+            {
+                ServiceLifetime.Singleton => MSServiceLifetime.Singleton,
+                ServiceLifetime.Scoped => MSServiceLifetime.Scoped,
+                ServiceLifetime.Transient => MSServiceLifetime.Transient,
+                _ => throw new ArgumentOutOfRangeException(nameof(serviceLifetime))
+            };
+
+        /// <inheritdoc/>
+        public override IGraphQLBuilder Configure<TOptions>(Action<TOptions, IServiceProvider>? action = null)
+        {
+            this.TryRegister(services => services.GetService<IOptions<TOptions>>()?.Value ?? new TOptions(), DI.ServiceLifetime.Singleton);
+            if (action != null)
+            {
+                this.Register<IConfigureOptions<TOptions>>(services => new ConfigureNamedOptions<TOptions>(Options.DefaultName, opt => action(opt, services)), ServiceLifetime.Singleton);
+            }
+
+            return this;
+        }
+
+        /// <inheritdoc/>
+        public override IGraphQLBuilder Register(Type serviceType, Func<IServiceProvider, object> implementationFactory, ServiceLifetime serviceLifetime)
+        {
+            if (serviceType == null)
+                throw new ArgumentNullException(nameof(serviceType));
+            if (implementationFactory == null)
+                throw new ArgumentNullException(nameof(implementationFactory));
+
+            _services.Add(new ServiceDescriptor(serviceType, implementationFactory, TranslateLifetime(serviceLifetime)));
+            return this;
+        }
+
+        /// <inheritdoc/>
+        public override IGraphQLBuilder Register(Type serviceType, Type implementationType, ServiceLifetime serviceLifetime)
+        {
+            if (serviceType == null)
+                throw new ArgumentNullException(nameof(serviceType));
+            if (implementationType == null)
+                throw new ArgumentNullException(nameof(implementationType));
+
+            _services.Add(new ServiceDescriptor(serviceType, implementationType, TranslateLifetime(serviceLifetime)));
+            return this;
+        }
+
+        /// <inheritdoc/>
+        public override IGraphQLBuilder Register(Type serviceType, object implementationInstance)
+        {
+            if (serviceType == null)
```

Comment:
Check this https://github.com/EasyNetQ/EasyNetQ/pull/1313 . What do you think ?

It will look like 
```
public override IGraphQLBuilder Register(Type serviceType, object implementationInstance)
{
            _services.Add(new ServiceDescriptor(serviceType.NotNull(), implementationInstance.NotNull()));
            return this;
}
```

I really do not like all this spaghetti code.

Reply:
Especially now with .NET 6 SDK we are ready for C# 10.

Reply:
I really dislike extension methods that exist only for type checking (and throw errors). Like DI, I think that methods are most clear when there is a dedicated place at the top of the method for input checks. 

Now it is not wholly bad. But the stack trace is likely inaccurate. Also it should not redefine the message for ArgumentNullException.

I agree it does simplify code in places like this. But I would not change any GraphQL.NET code to use `.NotNull()` or similar.

Reply:
I understand.

Reply:
See #2780